### PR TITLE
catalog-backend: fix sticky refresh state references

### PR DESCRIPTION
**Pull request by Rugvip**
Co-authored-by: Fredrik Adelöw <freben@gmail.com>

## Hey, I just made a Pull Request!

Replaces #15111 with this proposed fix. It's on top of the catalog storage refactor in #15143.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/catalog-backend/src/database/DefaultProcessingDatabase.ts [hunk 1]
```diff
@@ -357,13 +356,11 @@ export class DefaultProcessingDatabase implements ProcessingDatabase {
         this.options.logger.warn(
           `Detected conflicting entityRef ${entityRef} already referenced by ${conflictingKey} and now also ${locationKey}`,
         );
-        conflictingStateReferences.push(entityRef);
       }
     }
 
     // Replace all references for the originating entity or source and then create new ones
     await tx<DbRefreshStateReferencesRow>('refresh_state_references')
-      .whereNotIn('target_entity_ref', conflictingStateReferences)
```
Comment (Rugvip) [inline plugins/catalog-backend/src/database/DefaultProcessingDatabase.ts:366]: As far as we can till this behavior was incorrect, and we updated the test. The references from entity E1 to E2 should not stick around if E1 is no longer the producer of E2

File: plugins/catalog-backend/src/database/DefaultProviderDatabase.ts [hunk 2]
```diff
@@ -161,6 +161,11 @@ export class DefaultProviderDatabase implements ProviderDatabase {
             });
           }
 
+          await tx<DbRefreshStateReferencesRow>('refresh_state_references')
+            .where('target_entity_ref', entityRef)
+            .andWhere({ source_key: options.sourceKey })
+            .delete();
```
Comment (Rugvip) [inline plugins/catalog-backend/src/database/DefaultProviderDatabase.ts:167]: This is the fix for the original issue, which lines up the logic with the equivalent action in the processing database. I imagine we'll likely look to introduce a unique constraint here too though.

**Comment by freben** [comment]
Reassigned to @jhaals

**Review by jhaals — APPROVED** [review]

File: plugins/catalog-backend/src/database/DefaultProviderDatabase.ts (hunk 2 above)
Comment (webark) [reply plugins/catalog-backend/src/database/DefaultProviderDatabase.ts:167]: Will this handle cases where sourceKey is null?

Comment (Rugvip) [reply plugins/catalog-backend/src/database/DefaultProviderDatabase.ts:167]: This table has two different sets of rows, source_key + target, and source_ref + target. The source_key is for references from entity providers to entities, while the source_ref is references from one entity to another. This code only deals with the source_key + target rows, as it's only encapsulating the operations used by entity providers.

Comment (webark) [reply plugins/catalog-backend/src/database/DefaultProviderDatabase.ts:167]: sorry. I was thinking `source_entity_ref`. Nvm

**Review by alde — APPROVED** [review]

**Pull request merged**